iquip/apps/oven.py

```diff
@@ -0,0 +1,521 @@
+        """Adds the client object."""
+        client = self._client  # pylint: disable=protected-access
+        if client is None:
+            logger.error("Failed to get client.")
+            self.clientError.emit(ValueError("There is no client."))
+            return
+        try:
+            function(self, client, *args, **kwargs)
+        except (AttributeError, OSError, ValueError) as error:
+            logger.exception("Error occurred while running %s with the client.", function.__name__)
+            self.clientError.emit(error)
+            self._closeTarget()  # pylint: disable=protected-access
+    return wrapped
+
+
+class OvenManager(QObject):
+    """Manages the power supply unit RPC client for oven which lives in a dedicated thread.
+    
+    An instance of this class should be moved to a thread other than the main
+      GUI thread to prevent GUI from freezing.
+    Therefore, the private methods must not be called from the main thread.
+    Instead, use signals to communicate.
+
+    Signals:
+        connectionChanged(connected): The client connection status is changed,
+          with the connection status as True for connected, False for disconnected.
+        outputChanged(outputted): The output status is changed,
+          with the output status as True for turned on, False for turned off.
+        clientError(exception): An exception is occurred during client operation,
+          with the exception object.
+        [current, voltage]Reported([current, voltage]): The [current, voltage] is reported,
+          with its value.
+        See _signal() method for the other signal's.
+    """
+
+    connectionChanged = pyqtSignal(bool)
+    outputChanged = pyqtSignal(bool)
+    clientError = pyqtSignal(Exception)
+    currentReported = pyqtSignal(float)
+    voltageReported = pyqtSignal(float)
+
+    closeTarget = pyqtSignal()
+    openTarget = pyqtSignal(tuple)
+    readCurrent = pyqtSignal()
+    readVoltage = pyqtSignal()
+    setCurrent = pyqtSignal(float)
+    output = pyqtSignal(bool)
+
+    def __init__(self, parent: Optional[QObject] = None):
+        """Extended."""
+        super().__init__(parent=parent)
+        self._client: Optional[Client] = None
+        api = (
+            "closeTarget",
+            "openTarget",
+            "readCurrent",
+            "readVoltage",
+            "setCurrent",
+            "output",
+        )
+        for name in api:
+            signal = getattr(self, name)
+            method = getattr(self, f"_{name}")
+            signal.connect(method, type=Qt.QueuedConnection)
+
+    @pyqtSlot()
+    def _closeTarget(self):
+        """Closes the RPC client."""
+        if self._client is None:
+            logger.error("Failed to close target: RPC client does not exist.")
+            return
+        self._client.close_rpc()
+        self.connectionChanged.emit(False)
+
+    @pyqtSlot(tuple)
+    def _openTarget(self, info: RPCTargetInfo):
```

Comment:
How about unpacking `info` in the body to prevent confusing, for example,
```python
ip, port, target_name, target_channel = info
```
instead of using it as `info[3]`, etc.?